src/main/java/com/onelogin/saml/Response.java

```diff
@@ -219,31 +273,193 @@ public HashMap getAttributes() {
+	 *
+	 * @return true if the document passes.
+	 */
+	private boolean validateNumAssertions(){
+		NodeList assertionNodes = this.document.getElementsByTagNameNS("urn:oasis:names:tc:SAML:2.0:assertion", "Assertion");
+		if(assertionNodes != null && assertionNodes.getLength() == 1)
+			return true;
 		return false;
 	}
 
-	private void tagIdAttributes(Document xmlDoc) {
-		throw new UnsupportedOperationException("Not supported yet."); 
+	/**
+	 * Verifies that the document has the expected signed nodes.
+	 *
+	 * @return true if is valid
+	 */
+	private boolean validateSignedElements(ArrayList<String> signedElements){
+		if(signedElements.size() > 2){
+			return false;
+		}
+		Map<String, Integer> occurrences = new HashMap<String, Integer>();
+		for(String e:signedElements){
+			if(occurrences.containsKey(e)){
+				occurrences.put(e, occurrences.get(e).intValue() + 1);
+			}else{
+				occurrences.put(e, 1);
+			}
+		}
+
+		if((occurrences.containsKey("Response") && occurrences.get("Response") > 1) ||
+				(occurrences.containsKey("Assertion") && occurrences.get("Assertion") > 1) ||
+				!occurrences.containsKey("Response") && !occurrences.containsKey("Assertion")
+				) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Verifies that the document is still valid according Conditions Element.
+	 *
+	 * @return true if still valid
+	 */
+	private boolean validateTimestamps()
+	{
+		NodeList timestampNodes = document.getElementsByTagNameNS("*", "Conditions");
+		if (timestampNodes.getLength() != 0) {
+			for (int i = 0; i < timestampNodes.getLength(); i++) {
+				NamedNodeMap attrName = timestampNodes.item(i).getAttributes();
+				Node nbAttribute = attrName.getNamedItem("NotBefore");
+				Node naAttribute = attrName.getNamedItem("NotOnOrAfter");
+				Calendar now = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+				log.debug("now :"+ now.get(Calendar.HOUR_OF_DAY) + ":" + now.get(Calendar.MINUTE)+ ":" + now.get(Calendar.SECOND));
+				// validate NotOnOrAfter using UTC
+				if(naAttribute != null){						
+					final Calendar notOnOrAfterDate = javax.xml.bind.DatatypeConverter.parseDateTime(naAttribute.getNodeValue());
+					log.debug("notOnOrAfterDate :"+ notOnOrAfterDate.get(Calendar.HOUR_OF_DAY) + ":" + notOnOrAfterDate.get(Calendar.MINUTE)+ ":" + notOnOrAfterDate.get(Calendar.SECOND));
+					if(now.equals(notOnOrAfterDate) || now.after(notOnOrAfterDate)){
+						return false;
+					}
+				}
+				// validate NotBefore using UTC
+				if(nbAttribute != null){						
+					final Calendar notBeforeDate = javax.xml.bind.DatatypeConverter.parseDateTime(nbAttribute.getNodeValue());
+					log.debug("notBeforeDate :"+ notBeforeDate.get(Calendar.HOUR_OF_DAY) + ":" + notBeforeDate.get(Calendar.MINUTE)+ ":" + notBeforeDate.get(Calendar.SECOND));
+					if(now.before(notBeforeDate)){
+						return false;
+					}
+				}
+			}
+		}
+		return true;
 	}
 
 	public void setDestinationUrl(String urld){
```

Comment:
We never use this setDestinationUrl method on the response, and maybe we may try to calculate the currentUrl inside the toolkit.